File: SexyBiscuit.Engine/Core/Transform3D.cs
using Microsoft.Xna.Framework;

namespace SexyBiscuit.Engine.Core;

/// <summary>
/// 3D transform component — position (Vector3), rotation (Quaternion), scale (Vector3),
/// with full parent/child hierarchy and world/local space helpers.
/// </summary>
public sealed class Transform3D : Component
{
    // -------------------------------------------------------------------------
    // Local space
    // -------------------------------------------------------------------------
    private Vector3    _localPosition = Vector3.Zero;
    private Quaternion _localRotation = Quaternion.Identity;
    private Vector3    _localScale    = Vector3.One;

    public Vector3 LocalPosition
    {
        get => _localPosition;
        set { _localPosition = value; MarkDirty(); }
    }

    public Quaternion LocalRotation
    {
        get => _localRotation;
        set { _localRotation = value; MarkDirty(); }
    }

    public Vector3 LocalScale
    {
        get => _localScale;
        set { _localScale = value; MarkDirty(); }
    }

    public Vector3 LocalEulerAngles
    {
        get => QuaternionToEuler(_localRotation);
        set { _localRotation = EulerToQuaternion(value); MarkDirty(); }
    }

    // -------------------------------------------------------------------------
    // World space (lazy)
    // -------------------------------------------------------------------------
    private Vector3    _worldPosition;
    private Quaternion _worldRotation;
    private Vector3    _worldScale;
    private bool       _dirty = true;

    public Vector3 Position
    {
        get { Recalculate(); return _worldPosition; }
        set
        {
            // The world point must be expressed in the PARENT's space to become a local
            // offset. Inverting through this transform would be circular — it would use
            // the very position being assigned.
            _localPosition = _parent == null ? value : _parent.InverseTransformPoint(value);
            MarkDirty();
        }
    }

    public Quaternion Rotation
    {
        get { Recalculate(); return _worldRotation; }
        set
        {
            _localRotation = _parent == null
                ? value
                : Quaternion.Inverse(_parent.Rotation) * value;
            MarkDirty();
        }
    }

    public Vector3 Scale
    {
        get { Recalculate(); return _worldScale; }
        set
        {
            if (_parent == null)
                _localScale = value;
            else
            {
                var ps = _parent.Scale;
                _localScale = ps.X != 0 && ps.Y != 0 && ps.Z != 0
                    ? new Vector3(value.X / ps.X, value.Y / ps.Y, value.Z / ps.Z)
                    : value;
            }
            MarkDirty();
        }
    }

    public Vector3 EulerAngles
    {
        get => QuaternionToEuler(Rotation);
        set => Rotation = EulerToQuaternion(value);
    }

    // -------------------------------------------------------------------------
    // Direction vectors
    // -------------------------------------------------------------------------
    public Vector3 Forward => Vector3.Transform(Vector3.Forward, Rotation);
    public Vector3 Right   => Vector3.Transform(Vector3.Right,   Rotation);
    public Vector3 Up      => Vector3.Transform(Vector3.Up,      Rotation);

    // -------------------------------------------------------------------------
    // Hierarchy
    // -------------------------------------------------------------------------
    private Transform3D?       _parent;
    private List<Transform3D>  _children = new();

    public Transform3D? Parent => _parent;
    public IReadOnlyList<Transform3D> Children => _children;

    public void SetParent(Transform3D? newParent, bool keepWorldTransform = true)
    {
        if (_parent == newParent) return;

        Vector3    wp = keepWorldTransform ? Position : LocalPosition;
        Quaternion wr = keepWorldTransform ? Rotation : LocalRotation;
        Vector3    ws = keepWorldTransform ? Scale    : LocalScale;

        _parent?.RemoveChild(this);
        _parent = newParent;
        _parent?.AddChild(this);

        if (keepWorldTransform)
        {
            Position = wp;
            Rotation = wr;
            Scale    = ws;
        }

        MarkDirty();
    }

    private void AddChild(Transform3D c)    => _children.Add(c);
    private void RemoveChild(Transform3D c) => _children.Remove(c);

    // -------------------------------------------------------------------------
    // Matrix
    // -------------------------------------------------------------------------
    public Matrix GetLocalMatrix()
        => Matrix.CreateScale(_localScale)
         * Matrix.CreateFromQuaternion(_localRotation)
         * Matrix.CreateTranslation(_localPosition);

    public Matrix GetWorldMatrix()
    {
        Recalculate();
        return Matrix.CreateScale(_worldScale)
             * Matrix.CreateFromQuaternion(_worldRotation)
             * Matrix.CreateTranslation(_worldPosition);
    }

    // -------------------------------------------------------------------------
    // Space conversion
    // -------------------------------------------------------------------------
    public Vector3 TransformPoint(Vector3 localPoint)
    {
        Recalculate();
        return Vector3.Transform(localPoint * _worldScale, _worldRotation) + _worldPosition;
    }

    public Vector3 InverseTransformPoint(Vector3 worldPoint)
    {
        Recalculate();
        var delta = worldPoint - _worldPosition;
        var invRot = Quaternion.Inverse(_worldRotation);
        var local = Vector3.Transform(delta, invRot);
        var s = _worldScale;
        return s.X != 0 && s.Y != 0 && s.Z != 0
            ? new Vector3(local.X / s.X, local.Y / s.Y, local.Z / s.Z)
            : local;
    }

    public Vector3 TransformDirection(Vector3 localDir)
        => Vector3.Transform(localDir, Rotation);

    // -------------------------------------------------------------------------
    // Look-at
    // -------------------------------------------------------------------------
    public void LookAt(Vector3 target, Vector3? up = null)
    {
        var dir = target - Position;
        if (dir.LengthSquared() < 1e-6f) return;
        var matrix = Matrix.CreateWorld(Position, Vector3.Normalize(dir), up ?? Vector3.Up);
        Rotation = Quaternion.CreateFromRotationMatrix(matrix);
    }

    // -------------------------------------------------------------------------
    // Recalculation
    // -------------------------------------------------------------------------
    private void Recalculate()
    {
        if (!_dirty) return;
        _dirty = false;

        if (_parent == null)
        {
            _worldPosition = _localPosition;
            _worldRotation = _localRotation;
            _worldScale    = _localScale;
        }
        else
        {
            _worldScale    = _parent.Scale * _localScale;
            _worldRotation = _parent.Rotation * _localRotation;
            _worldPosition = Vector3.Transform(_localPosition * _parent.Scale, _parent.Rotation)
                           + _parent.Position;
        }

    }

    /// <summary>
    /// Marks this transform and its whole subtree as needing recalculation.
    /// </summary>
    /// <remarks>
    /// Every local mutation goes through here rather than setting <c>_dirty</c> alone.
    /// Marking only self was not enough: a descendant whose own flag was already clear
    /// returned its cached world transform without ever consulting the ancestor that
    /// moved, so moving a root left the rest of the chain behind. Recursion stops at a
    /// subtree that is already dirty, so a burst of edits in one frame costs one walk.
    /// </remarks>
    private void MarkDirty()
    {
        if (_dirty) return;
        _dirty = true;

        foreach (var child in _children)
            child.MarkDirty();
    }

    // -------------------------------------------------------------------------
    // Euler conversions
    // -------------------------------------------------------------------------
    /// <summary>Builds a rotation from Euler angles in degrees: X pitch, Y yaw, Z roll.</summary>
    public static Quaternion EulerToQuaternion(Vector3 eulerDegrees)
    {
        float rx = MathHelper.ToRadians(eulerDegrees.X);
        float ry = MathHelper.ToRadians(eulerDegrees.Y);
        float rz = MathHelper.ToRadians(eulerDegrees.Z);
        return Quaternion.CreateFromYawPitchRoll(ry, rx, rz);
    }

    /// <summary>The inverse of <see cref="EulerToQuaternion"/>: pitch, yaw and roll in degrees.</summary>
    public static Vector3 QuaternionToEuler(Quaternion q)
    {
        float sinr = 2f * (q.W * q.X + q.Y * q.Z);
        float cosr = 1f - 2f * (q.X * q.X + q.Y * q.Y);
        float pitch = MathF.Atan2(sinr, cosr);

        float sinp = 2f * (q.W * q.Y - q.Z * q.X);
        float yaw = MathF.Abs(sinp) >= 1f ? MathF.CopySign(MathF.PI / 2f, sinp) : MathF.Asin(sinp);

        float siny = 2f * (q.W * q.Z + q.X * q.Y);
        float cosy = 1f - 2f * (q.Y * q.Y + q.Z * q.Z);
        float roll = MathF.Atan2(siny, cosy);

        return new Vector3(
            MathHelper.ToDegrees(pitch),
            MathHelper.ToDegrees(yaw),
            MathHelper.ToDegrees(roll));
    }
}

File: html5/src/core/Transform3D.js
// -----------------------------------------------------------------------------
// Transform3D — position, rotation and scale in 3D.
//
// Not added automatically: adding one is how a 2D actor becomes a 3D one, which
// is the same rule the C# engine follows and why a scene file only carries
// `position3`/`rotation3`/`scale3` for actors that have it.
//
// Euler angles are degrees in pitch, yaw, roll order, and the conversion routes
// through the same yaw-pitch-roll construction the C# `Transform3D` uses, so a
// rotation authored in either engine reads identically in the other.
// -----------------------------------------------------------------------------

import { Component } from './Component.js';
import { Vector3, Quaternion, Matrix4 } from '../math/index.js';

/** Position, rotation and scale in 3D, with a parent/child hierarchy. */
export class Transform3D extends Component {
    static schema = {};

    constructor() {
        super();
        this._localPosition = new Vector3(0, 0, 0);
        this._localRotation = Quaternion.identity;
        this._localScale = new Vector3(1, 1, 1);

        this._worldPosition = new Vector3(0, 0, 0);
        this._worldRotation = Quaternion.identity;
        this._worldScale = new Vector3(1, 1, 1);
        this._dirty = true;

        this._parent = null;
        this._children = [];
    }

    // ---- Local space ---------------------------------------------------------

    get localPosition() { return this._localPosition; }
    set localPosition(value) { this._localPosition = Vector3.from(value); this.markDirty(); }

    get localRotation() { return this._localRotation; }
    set localRotation(value) { this._localRotation = Quaternion.from(value); this.markDirty(); }

    get localScale() { return this._localScale; }
    set localScale(value) { this._localScale = Vector3.from(value); this.markDirty(); }

    /** Local rotation as pitch, yaw and roll in degrees. */
    get localEulerAngles() { return Quaternion.toEuler(this._localRotation); }
    set localEulerAngles(value) { this._localRotation = Quaternion.fromEuler(value); this.markDirty(); }

    // ---- World space ---------------------------------------------------------

    get position() { this._recalculate(); return this._worldPosition; }
    set position(value) {
        const world = Vector3.from(value);
        this._localPosition = this._parent ? this._parent.inverseTransformPoint(world) : world;
        this.markDirty();
    }

    get rotation() { this._recalculate(); return this._worldRotation; }
    set rotation(value) {
        const world = Quaternion.from(value);
        this._localRotation = this._parent
            ? Quaternion.multiply(Quaternion.inverse(this._parent.rotation), world)
            : world;
        this.markDirty();
    }

    get scale() { this._recalculate(); return this._worldScale; }
    set scale(value) {
        const world = Vector3.from(value);
        if (!this._parent) {
            this._localScale = world;
        } else {
            const ps = this._parent.scale;
            this._localScale = (ps.x !== 0 && ps.y !== 0 && ps.z !== 0)
                ? new Vector3(world.x / ps.x, world.y / ps.y, world.z / ps.z)
                : world;
        }
        this.markDirty();
    }

    /** World rotation as pitch, yaw and roll in degrees. */
    get eulerAngles() { return Quaternion.toEuler(this.rotation); }
    set eulerAngles(value) { this.rotation = Quaternion.fromEuler(value); }

    // ---- Convenience accessors ----------------------------------------------

    get x() { return this.position.x; }
    set x(v) { const p = this.position; this.position = new Vector3(v, p.y, p.z); }

    get y() { return this.position.y; }
    set y(v) { const p = this.position; this.position = new Vector3(p.x, v, p.z); }

    get z() { return this.position.z; }
    set z(v) { const p = this.position; this.position = new Vector3(p.x, p.y, v); }

    // ---- Direction vectors ---------------------------------------------------
    // Forward is -Z, matching MonoGame, so ported camera code stays correct.

    get forward() { return Vector3.transform(Vector3.forward, this.rotation); }
    get right()   { return Vector3.transform(Vector3.right, this.rotation); }
    get up()      { return Vector3.transform(Vector3.up, this.rotation); }

    // ---- Hierarchy -----------------------------------------------------------

    get parent() { return this._parent; }
    get children() { return this._children; }

    setParent(newParent, keepWorldTransform = true) {
        if (this._parent === newParent) return;

        const wp = keepWorldTransform ? this.position.clone() : this.localPosition.clone();
        const wr = keepWorldTransform ? this.rotation.clone() : this.localRotation.clone();
        const ws = keepWorldTransform ? this.scale.clone() : this.localScale.clone();

        if (this._parent) {
            const i = this._parent._children.indexOf(this);
            if (i >= 0) this._parent._children.splice(i, 1);
        }

        this._parent = newParent ?? null;
        if (this._parent) this._parent._children.push(this);

        if (keepWorldTransform) {
            this.position = wp;
            this.rotation = wr;
            this.scale = ws;
        }

        this.markDirty();
    }

    // ---- Matrices ------------------------------------------------------------

    getLocalMatrix() {
        return Matrix4.compose(this._localPosition, this._localRotation, this._localScale);
    }

    getWorldMatrix() {
        this._recalculate();
        return Matrix4.compose(this._worldPosition, this._worldRotation, this._worldScale);
    }

    // ---- Space conversion ----------------------------------------------------

    transformPoint(localPoint) {
        const p = Vector3.from(localPoint);
        const scaled = Vector3.multiply(p, this.scale);
        return Vector3.add(Vector3.transform(scaled, this.rotation), this.position);
    }

    inverseTransformPoint(worldPoint) {
        const p = Vector3.subtract(Vector3.from(worldPoint), this.position);
        const unrotated = Vector3.transform(p, Quaternion.inverse(this.rotation));
        const s = this.scale;
        return (s.x !== 0 && s.y !== 0 && s.z !== 0)
            ? new Vector3(unrotated.x / s.x, unrotated.y / s.y, unrotated.z / s.z)
            : unrotated;
    }

    /** Rotates a direction into world space, ignoring position and scale. */
    transformDirection(localDirection) {
        return Vector3.transform(Vector3.from(localDirection), this.rotation);
    }

    // ---- Utility -------------------------------------------------------------

    /** Rotates so the local forward axis points at `target`. */
    lookAt(target, up = Vector3.up) {
        const dir = Vector3.subtract(Vector3.from(target), this.position);
        if (dir.lengthSquared < 1e-12) return;
        this.rotation = Quaternion.lookRotation(dir.normalize(), Vector3.from(up));
    }

    /** Moves by a world-space delta. */
    translate(delta) {
        this.position = Vector3.add(this.position, Vector3.from(delta));
    }

    distanceTo(other) {
        const b = other?.position ?? Vector3.from(other);
        return Vector3.distance(this.position, b);
    }

    // ---- Internals -----------------------------------------------------------

    _recalculate() {
        if (!this._dirty) return;
        this._dirty = false;

        if (!this._parent) {
            this._worldPosition = this._localPosition.clone();
            this._worldRotation = this._localRotation.clone();
            this._worldScale = this._localScale.clone();
            return;
        }

        const pPos = this._parent.position;
        const pRot = this._parent.rotation;
        const pScl = this._parent.scale;

        this._worldScale = Vector3.multiply(pScl, this._localScale);
        this._worldRotation = Quaternion.multiply(pRot, this._localRotation);
        this._worldPosition = Vector3.add(
            Vector3.transform(Vector3.multiply(this._localPosition, pScl), pRot),
            pPos);
    }

    markDirty() {
        if (this._dirty) return;
        this._dirty = true;
        for (const child of this._children) child.markDirty();
    }

    /** Builds a rotation from pitch, yaw and roll in degrees. */
    static eulerToQuaternion(eulerDegrees) { return Quaternion.fromEuler(eulerDegrees); }

    /** The inverse: pitch, yaw and roll in degrees. */
    static quaternionToEuler(q) { return Quaternion.toEuler(q); }

    toString() {
        return `Transform3D(pos=${this._localPosition}, euler=${this.localEulerAngles})`;
    }
}
